### PR TITLE
karan solves excercise 1

### PR DESCRIPTION
**Pull request by zontyp**
added new file with my solution 
so that other users can review and comment as required.

### CODE REVIEW

File: app/Main_karan_solves.hs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+--ghci >1234 `div` 10000
+--0
+myReverse :: [Integer] -> [Integer]
+myReverse []  = []
+myReverse xs = myReverseAcc xs []
+  where 
+    myReverseAcc (x:xs) acc = myReverseAcc xs (x:acc)
+    myReverseAcc [] acc = acc
+
+doubleAlts :: [Integer] -> [Integer]
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs:42]: in your repl you can call this function with some argument.
if you do `doubleAlts [1,2,3,4]` what do you get ?
try REPL driven development.

File: app/Main_karan_solves.hs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+--ghci >1234 `div` 10
+--123
+--ghci >1234 `div` 100
+--12
+--ghci >1234 `div` 1000
+--1
+--ghci >1234 `div` 10000
+--0
+myReverse :: [Integer] -> [Integer]
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs:35]: 👍

File: app/Main_karan_solves.hs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+--ghci >1234 `rem` 1 0000  `div` 1000
+--1
+--ghci >1234 `rem` 1000 `div` 100
+--2
+--ghci >1234 `rem` 100 `div` 10
+--3
+--ghci >1234 `rem` 10 `div` 1
+--4
+toDigits :: Integer -> [Integer]
+toDigits cardNumber
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs]: Will `toDigits` pass the examples given in http://www.cis.upenn.edu/%7Ecis194/spring13/hw/01-intro.pdf ?

```
Example:
toDigits 1234 == [1,2,3,4]
Example:
toDigitsRev 1234 == [4,3,2,1]
Example:
toDigits 0 == []
Example:
toDigits (-17) == []
```

File: app/Main_karan_solves.hs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    ndigits = numdigits cardNumber 1
+
+-- You need to map over every element of the list..
+-- But you can use recursion for it as well..
+-- Otherwise you can write to the top of file after `module ..`
+-- import qualified Data.List as DL
+-- and use DL.map..
+-- but remember it is all about writing Recursion at this point..
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther =  myReverse . doubleAlts . myReverse
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs:90]: please use recursion, just like you used in `myReverseAcc`

File: app/Main_karan_solves.hs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+sumDigits :: [Integer] -> Integer
+sumDigits [] = 0
+sumDigits (x:xs) = x + sumDigits xs
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs:94]: 👍

File: app/Main_karan_solves.hs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+validate :: Integer -> Bool
+validate cardNumber 
+  | mychecksum `rem` 10 == 0 = True
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs]: `guard` patterns look a little overkill here..
using 

```
let mychecksum = ...
in mychecksum `rem` 10 == 0
```

would be better.

File: app/Main_karan_solves.hs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  | otherwise = False 
+      where mychecksum = sumDigits $ doubleEveryOther $ toDigits $ cardNumber
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs]: just a `tip` : 
`sumDigits . doubleEveryOther . toDigits $ cardNumber` will also work..
we will discuss how it works in next session :)

**Review by ashishnegi — COMMENTED** [review]
Overall this is satisfactory work.. nice job.
just go through the comments and try to rectify the errors.

File: app/Main_karan_solves.hs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+--1
+--ghci >1234 `rem` 1000 `div` 100
+--2
+--ghci >1234 `rem` 100 `div` 10
+--3
+--ghci >1234 `rem` 10 `div` 1
+--4
+toDigits :: Integer -> [Integer]
+toDigits cardNumber
+  | cardNumber == 0 = []
```
Comment (ashishnegi) [inline app/Main_karan_solves.hs]: i think you forgot to remove line no `69 and 70`.
either your function body uses guards or it does not.

File: app/Main_karan_solves.hs (hunk 1 above)
Comment (zontyp) [reply app/Main_karan_solves.hs:42]: i get 
[1,4,3,8]
 doublealts will double only alternate elements of a list.

i shall add comment for this function in code.

File: app/Main_karan_solves.hs (hunk 3 above)
Comment (zontyp) [reply app/Main_karan_solves.hs]: yes .
this is working.

File: app/Main_karan_solves.hs (hunk 8 above)
Comment (zontyp) [reply app/Main_karan_solves.hs]: thanks bunch for this comment.
this mistake exists and willbe modified.

File: app/Main_karan_solves.hs (hunk 7 above)
Comment (zontyp) [reply app/Main_karan_solves.hs]: agree.

any tradeoffs between $ and (.)   ?

File: app/Main_karan_solves.hs (hunk 6 above)
Comment (zontyp) [reply app/Main_karan_solves.hs]: done.
thanks.

**Pull request merged**